plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtFreezeThawVMCommandWrapper.java

```diff
@@ -0,0 +1,104 @@
+
+        try {
+            final LibvirtUtilitiesHelper libvirtUtilitiesHelper = serverResource.getLibvirtUtilitiesHelper();
+            Connect connect = libvirtUtilitiesHelper.getConnection();
+            domain = serverResource.getDomain(connect, vmName);
+            if (domain == null) {
+                return new FreezeThawVMAnswer(command, false, String.format("Failed to %s due to %s was not found",
+                        command.getOption(), vmName));
+            }
+            DomainState domainState = domain.getInfo().state ;
+            if (domainState != DomainState.VIR_DOMAIN_RUNNING) {
+                return new FreezeThawVMAnswer(command, false,
+                        String.format("%s of VM failed due to vm %s is in %s state", command.getOption(),
+                                vmName, domainState));
+            }
+
+            String result = getResultOfQemuCommand(command.getOption(), domain);
+            s_logger.debug(String.format("Result of %s command is %s", command.getOption(), result));
+            if (result == null || (result.startsWith("error"))) {
+                return new FreezeThawVMAnswer(command, false, String.format("Failed to %s vm %s due to result status is: %s",
+                        command.getOption(), vmName, result));
+            }
+            String status = getResultOfQemuCommand(FreezeThawVMCommand.STATUS, domain);
+            s_logger.debug(String.format("Status of %s command is %s", command.getOption(), status));
+            if (status != null && new JsonParser().parse(status).isJsonObject()) {
+                String statusResult = new JsonParser().parse(status).getAsJsonObject().get("return").getAsString();
+                if (statusResult.equals(command.getOption())) {
+                    return new FreezeThawVMAnswer(command, true, String.format("%s of VM - %s is successful", command.getOption(), vmName));
+                }
+            }
+            return new FreezeThawVMAnswer(command, false, String.format("Failed to %s vm %s due to result status is: %s",
+                    command.getOption(), vmName, status));
+        } catch (LibvirtException libvirtException) {
+            return new FreezeThawVMAnswer(command, false,  String.format("Failed to %s VM - %s due to %s",
+                    command.getOption(), vmName, libvirtException.getMessage()));
+        }finally {
+            if (domain != null) {
+                try {
+                    domain.free();
+                } catch (LibvirtException e) {
+                    s_logger.trace("Ingore error ", e);
+                }
+            }
+        }
+    }
+
+    private String getResultOfQemuCommand(String cmd, Domain domain) throws LibvirtException {
+        String result = null;
+        if (cmd.equals(FreezeThawVMCommand.FREEZE)) {
+            result = domain.qemuAgentCommand(new Gson().toJson(QemuCommand.executeQemuCommand(QemuCommand.AGENT_FREEZE, null)).toString(), 10, 0);
```

Comment:
I think readability could be increased on these 3 calls by making the `QemuCommand.executeQemuCommand` (or renamed) method return a string instead of a map.

Also, these 3 calls could be reduced to only 1, by defining a collection that returns the agent "guest-fsfreeze-*" command based on the `cmd`

Reply:
@nvazquez, I prefer to leave it like this. In the feature, we could reuse this for commands that accept arguments, and for me, it will be more readable than assembling different strings.

Reply:
@slavkap ok, since qemuAgentCommand accepts a string as a first parameter I would be +1 on having a helper class or similar that will retrieve the string in the expected format based on the params, but not a blocker